### PR TITLE
Added "Core" layer for solution

### PR DESCRIPTION
**Pull request by s-samusik**

### CODE REVIEW

File: DatasetMarketplace.Domain/Entities/OrderCard.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace DatasetMarketplace.Domain.Entities
+{
+    public class OrderCard
+    {
+        public Guid Id { get; set; }
+        public Guid DatasetId { get; set; }
+        public Guid BuyerId { get; set; }
+        public OrderStatus OrderStatus { get; set; }
+        public PaymentType PaymentType { get; set; }
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Domain/Entities/OrderCard.cs:12]: Mark as nullable

File: DatasetMarketplace.Data/DatasetMarketplace.Data.csproj [hunk 2]
```diff
@@ -2,15 +2,16 @@
 
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
-    <RootNamespace>_03.DatasetMarketplace.Data</RootNamespace>
+    <RootNamespace>DatasetMarketplace.Data</RootNamespace>
```
[the remark this answers is not in the document]
Comment (s-samusik) [reply DatasetMarketplace.Data/DatasetMarketplace.Data.csproj:5]: I'm going to use this project for repositories (interfaces and implementation). Maybe name of project is incorrect.
I remember you suggested me work repositories in once project, don't sesparate interfaces and implementation on different layers. 
Or maybe i need to work with repositories in "Persistence" layer and rename "Persistence" to "DataAccess" or better "Infrastructure"? 
Also i have variant to move interfaces of repositories in "Persistence" and implement it in "Data" layer and renamed "Data" to "Infrastructure".
I don't know which variant is right becouse there is have no only right variant here.

File: DatasetMarketplace.Api/Startup.cs [hunk 3]
```diff
@@ -21,13 +17,14 @@ public Startup(IConfiguration configuration)
 
         public IConfiguration Configuration { get; }
 
-        // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddControllers();
+
+            services.AddDbContext<DatasetMarketplaceContext>(options =>
+                options.UseNpgsql(Configuration["DatasetMarketlpace:PostgreSQL"]));
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Api/Startup.cs:25]: Return this change back.

And return change in json file, but just don't push your credentials. Do like this:

> `"ConnectionStrings": {`
>     `"PostgreSQL": "<connection string>"`
>   `},`

File: DatasetMarketplace.Domain/Entities/User.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    public class User : BaseEntity
+    {
+        public string FirstName { get; set; }
+        public string LastName { get; set; }
+        public string Email { get; set; }
+        public string PhoneNumber { get; set; }
+        public string PasswordHash { get; set; }
+        public string PasswordSalt { get; set; }
+
+        public virtual ICollection<Role> UserRoles { get; set; } = new HashSet<Role>();
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Domain/Entities/User.cs:14]:
```suggestion
        public virtual ICollection<Role> Roles { get; set; } = new HashSet<Role>();
```

File: DatasetMarketplace.Persistence/Configurations/UserRoleConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+﻿using DatasetMarketplace.Domain.Entities;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace DatasetMarketplace.Persistence.Configurations
+{
+    public class UserRoleConfiguration : IEntityTypeConfiguration<Role>
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Persistence/Configurations/UserRoleConfiguration.cs:7]:
```suggestion
    public class RoleConfiguration : IEntityTypeConfiguration<Role>
```

File: DatasetMarketplace.Persistence/DatasetMarketplaceContext.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        {
+        }
+
+        public virtual DbSet<Company> Companies { get; set; }
+        public virtual DbSet<Dataset> Datasets { get; set; }
+        public virtual DbSet<Deal> Deals { get; set; }
+        public virtual DbSet<OrderCard> OrderCards { get; set; }
+        public virtual DbSet<PlacementCard> PlacementCards { get; set; }
+        public virtual DbSet<User> Users { get; set; }
+        public virtual DbSet<Role> UserRoles { get; set; }
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Persistence/DatasetMarketplaceContext.cs:21]:
```suggestion
        public virtual DbSet<Role> Roles { get; set; }
```

File: DatasetMarketplace.Persistence/Configurations/UserConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace DatasetMarketplace.Persistence.Configurations
+{
+    public class UserConfiguration : IEntityTypeConfiguration<User>
+    {
+        public void Configure(EntityTypeBuilder<User> builder)
+        {
+            builder.HasKey(x => x.Id);
+
+            builder.Property(x => x.FirstName);
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Persistence/Configurations/UserConfiguration.cs:13]: Do we really need code lines like this. We do nothing with property?

File: DatasetMarketplace.Persistence/Configurations/PlacementCardConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+{
+    public class PlacementCardConfiguration : IEntityTypeConfiguration<PlacementCard>
+    {
+        public void Configure(EntityTypeBuilder<PlacementCard> builder)
+        {
+            builder.HasKey(x => x.Id);
+
+            builder.Property(x => x.OwnerId);
+
+            builder.Property(x => x.DatasetId);
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Persistence/Configurations/PlacementCardConfiguration.cs:15]: The same...

File: DatasetMarketplace.Persistence/Configurations/DatasetConfiguration.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    public class DatasetConfiguration : IEntityTypeConfiguration<Dataset>
+    {
+        public void Configure(EntityTypeBuilder<Dataset> builder)
+        {
+            builder.HasKey(x => x.Id);
+
+            builder.Property(x => x.Title)
+                .IsRequired();
+
+            builder.Property(x => x.Description);
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Persistence/Configurations/DatasetConfiguration.cs:16]: The same...

File: DatasetMarketplace.Domain/Enums/CardStatus.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿namespace DatasetMarketplace.Domain.Enums
+{
+    public enum CardStatus
+    {
+        Pending = 0,
+        Processing,
+        Canceled,
+        Failed,
+        Successed
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Domain/Enums/CardStatus.cs:9]: Values must be explicitly specified for all elements of the enums.

File: DatasetMarketplace.Domain/Entities/Role.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Collections.Generic;
+using DatasetMarketplace.Domain.Enums;
+
+namespace DatasetMarketplace.Domain.Entities
+{
+    public class Role : BaseEntity
+    {
+        public RoleType Name { get; set; }
```
Comment (DzmitryBratchuk) [inline DatasetMarketplace.Domain/Entities/Role.cs:8]: public RoleType RoleType { get; set; } or public RoleType Type { get; set; }

**Review by DzmitryBratchuk — CHANGES_REQUESTED** [review]

File: DatasetMarketplace.Data/DatasetMarketplace.Data.csproj (hunk 2 above)
[the remark this answers is not in the document]
Comment (DzmitryBratchuk) [reply DatasetMarketplace.Data/DatasetMarketplace.Data.csproj:5]: > Or maybe i need to work with repositories in "Persistence" layer and rename "Persistence" to "DataAccess" or better "Infrastructure"?

I prefere this one. Work with repositories in "Persistence" layer and rename "Persistence" to "Data" or "DataAccess"

**Review by DzmitryBratchuk — APPROVED** [review]

**Pull request merged**